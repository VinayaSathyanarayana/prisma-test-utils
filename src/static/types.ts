import { Dictionary } from 'lodash'

/**
 * Types used to describe the seed models.
 */
export type SeedFunction<PhotonType, GeneratedSeedModels extends SeedModels> = (
  options: SeedOptions<PhotonType, GeneratedSeedModels>,
) => Promise<object[]>

export type SeedModelsDefinition<GeneratedSeedModels extends SeedModels> = (
  kit: SeedKit,
) => GeneratedSeedModels

export interface SeedKit {
  faker: Chance.Chance
}

export type SeedModels = {
  '*': { amount: number }
  [type: string]: SeedModel
}

export type SeedModel = {
  amount?: number
  factory?: Dictionary<
    | SeedModelFieldDefinition
    | (() => SeedModelFieldDefinition)
    | SeedModelFieldRelationConstraint
  >
}

export type ID = string | number

export type SeedModelFieldDefinition = ID | string | number | boolean

export type SeedModelFieldRelationConstraint = {
  min?: number
  max?: number
}

export type SeedOptions<PhotonType, GeneratedSeedModel extends SeedModels> = {
  client: PhotonType
  models: SeedModelsDefinition<GeneratedSeedModel>
  seed?: number
  // persist?: boolean
}

/**
 * Types used to describe the Pool.
 */

export type DBInstance = {
  url: string
  cwd: string
  datamodel: string
}

export abstract class Pool {
  abstract async getDBInstance(): Promise<DBInstance>
  abstract async releaseDBInstance(db: DBInstance): Promise<void>
  abstract async run<T>(fn: (db: DBInstance) => Promise<T>): Promise<T>
  abstract async drain(): Promise<void>
}
