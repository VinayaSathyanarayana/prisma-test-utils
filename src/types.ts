import Field from '@prisma/dmmf/dist/Field'
import Model from '@prisma/dmmf/dist/Model'
import { Dictionary } from 'lodash'

/**
 * FakerBag is a set of tools that people can use to define how
 * their model is generated.
 */
export interface FakerBag {
  faker: Faker.FakerStatic
}

export type RelationConstraint = {
  min?: number
  max?: number
}

/**
 * FakerModel represents the collection of all explicit faking model definitions
 * used in Prisma Faker.
 */
export type FakerSchema = Dictionary<FixtureDefinition>

export type FixtureDefinition = {
  amount?: number
  factory?: Dictionary<FixtureFieldDefinition | (() => FixtureFieldDefinition)>
}

export type FixtureFieldDefinition = string | number | RelationConstraint

/**
 * Prisma Faker uses intermediate step between model-faker definition
 * and database seeding. Fixture is a virtual data unit used to describe
 * future data and calculate relations.
 */
export interface Order {
  model: Model
  amount: number
  relations: { [field: string]: number }
}

export interface Step {
  order: number // the creation order of a step, starts with 0
  model: Model
  amount: number // number of instances created in this step
  runningNumber: number // specifies the total number of all instances
  relations: { [field: string]: number }
}

/**
 * Represents the virtual unit.
 */
export interface Fixture {
  order: number // starts with 0
  id: string
  model: Model
}
