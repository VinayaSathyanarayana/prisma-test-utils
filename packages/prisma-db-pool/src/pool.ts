import { LiftEngine, DataSource } from '@prisma/lift'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import pify from 'pify'
import { withDefault } from './utils'

const writeFile = pify(fs.writeFile)
const makeDir = pify(fs.mkdir)

export type PoolDefinition = {
  dmmf: DMMF.Document
  pool: {
    min?: number
    max: number
  }
  cwd?: string
}

export type DBInstance = {
  cwd: string
  datamodel: string
}

export class Pool {
  private dmmf: DMMF.Document
  private cwd: string
  private dbs: {
    booting: string[]
    idle: DBInstance[]
    busy: DBInstance[]
  }
  private waiters: Waiter[]
  private capacity: number

  constructor(definition: PoolDefinition) {
    this.dmmf = definition.dmmf
    this.cwd = withDefault(process.cwd(), definition.cwd)
    this.dbs = {
      booting: [],
      idle: [],
      busy: [],
    }
    this.waiters = []
    this.capacity = definition.pool.max

    /* Start init */

    for (let index = 0; index < withDefault(0, definition.pool.min); index++) {
      this.createDBInstance().then(db => {
        this.dbs.idle = [...this.dbs.idle, db]
      })
    }
  }

  /**
   * Creates a new DB instances and lifts a migration.
   */
  private async createDBInstance(): Promise<DBInstance> {
    try {
      /* Constants */

      const id = Math.random()
        .toString(36)
        .slice(2)
      const tmpDir = os.tmpdir()
      const datasources: DataSource[] = [
        {
          name: 'db',
          connectorType: 'sqlite',
          url: `file:${tmpDir}`,
          config: {},
        },
      ]
      const dbFolder = path.join(tmpDir, './db/')

      /* Occupy resource pool. */
      this.dbs.booting = [...this.dbs.booting, id]

      /* Migrate Datamodel */
      const lift = new LiftEngine({
        projectDir: this.cwd,
      })

      const { datamodel } = await lift.convertDmmfToDml({
        dmmf: JSON.stringify(this.dmmf.datamodel),
        dataSources: datasources,
      })

      /* Release resource pool. */
      this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

      return {
        cwd: tmpDir,
        datamodel: datamodel,
      }
    } catch (err) {
      throw err
    }
  }

  /**
   * Returns a Photon instance from the pool once made available, and cleans
   * and populates the instance if necessary.
   *
   * The returned Photon instance includes the db identifier used for
   * the database releasing.
   *
   * @param opts
   */
  async getDBInstance(): Promise<DBInstance> {
    /**
     * Check whether any of the instances is available and creates a new one
     * if the pool limit is not yet reached, otherwise waits for the available
     * instance.
     */
    if (this.dbs.idle.length > 0) {
      /* If available take a resource and return it right away. */
      const [db, ...remaining] = this.dbs.idle
      this.dbs.idle = remaining
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else if (
      this.dbs.idle.length + this.dbs.busy.length + this.dbs.booting.length <
      this.capacity
    ) {
      /* If full capacity is not yet reached create new instance. */
      const db = await this.createDBInstance()
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else {
      /* Add to the line. */
      const waiter = new Waiter()
      this.waiters = [...this.waiters, waiter]

      return waiter.wait()
    }
  }

  /**
   * Makes the instance available.
   */
  async releaseDBInstance(db: DBInstance): Promise<void> {
    /**
     * Find the busy instance, remove that instance and give
     * it to the next in line.
     */
    const instance = this.dbs.busy.find(bdb => bdb.cwd === db.cwd)
    this.dbs.busy.filter(bdb => bdb.cwd !== db.cwd)

    /* Add to the next in line or make idle. */
    if (this.waiters.length > 0) {
      const [waiter, ...remainingWaiters] = this.waiters
      this.waiters = remainingWaiters
      waiter.allocate(instance)
    } else {
      this.dbs.idle = [...this.dbs.idle, instance]
    }
  }

  /**
   * Drains the pool by deleting all instances.
   */
  async drain(): Promise<void> {
    // TODO
  }
}

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
