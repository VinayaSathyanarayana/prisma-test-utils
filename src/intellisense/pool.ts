import { GeneratorOptions } from '@prisma/generator-helper'

/**
 * Returns a dedicated function name matching the specified datasource.
 *
 * @param dmmf
 */
export function generatePoolType(
  options: GeneratorOptions,
): 'getPostgreSQLPool' | 'getMySQLPool' | 'getSQLitePool' {
  const [datasource] = options.datasources

  /* istanbul ignore if */
  if (!datasource) {
    throw new Error(`No defined datasource!`)
  }

  switch (datasource.connectorType) {
    case 'sqlite': {
      return 'getSQLitePool'
    }
    case 'mysql': {
      return 'getMySQLPool'
    }
    case 'postgresql': {
      return 'getPostgreSQLPool'
    }
    /* istanbul ignore next */
    default: {
      throw new Error(`Unknown datasource: ${datasource.connectorType}`)
    }
  }
}
