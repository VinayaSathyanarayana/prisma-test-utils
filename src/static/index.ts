export {
  getMySQLPool,
  getPostgreSQLPool,
  getSQLitePool,
  MySQLPoolOptions,
  PostgreSQLPoolOptions,
  SQLitePoolOptions,
} from './pool'
export { getSeed } from './seed'
export * from './types'
