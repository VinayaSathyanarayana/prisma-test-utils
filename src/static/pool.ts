import { LiftEngine, DataSource } from '@prisma/lift'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import * as os from 'os'
import * as path from 'path'
import { withDefault } from '../utils'

export type PoolOptions = {
  pool: {
    max?: number
  }
}

export type DBInstance = {
  url: string
  cwd: string
  datamodel: string
}

export default (dmmf: DMMF.Document) =>
  class Pool {
    private dbs: {
      booting: string[]
      idle: DBInstance[]
      busy: DBInstance[]
    }
    private waiters: Waiter[]
    private capacity: number

    constructor(definition: PoolOptions) {
      this.dbs = {
        booting: [],
        idle: [],
        busy: [],
      }
      this.waiters = []
      this.capacity = withDefault(Infinity, definition.pool.max)
    }

    /**
     * Creates a new DB instances and lifts a migration.
     */
    private async createDBInstance(): Promise<void> {
      try {
        /* Constants */

        const id = Math.random()
          .toString(36)
          .slice(2)
        const tmpDir = path.join(os.tmpdir(), `prisma-pool-${id}`)
        const dbFile = path.join(tmpDir, './db.db')
        const datasources: DataSource[] = [
          {
            name: 'db',
            connectorType: 'sqlite',
            url: `file:${dbFile}`,
            config: {},
          },
        ]

        /* Occupy resource pool. */
        this.dbs.booting = [...this.dbs.booting, id]

        /* Migrate Datamodel. */
        const lift = new LiftEngine({
          projectDir: tmpDir,
        })

        const datamodelDmmf = {
          enums: [],
          models: [],
          ...dmmf.datamodel,
        }

        const { datamodel } = await lift.convertDmmfToDml({
          dmmf: JSON.stringify(datamodelDmmf),
          config: { datasources, generators: [] },
        })

        const {
          datamodelSteps,
          errors: stepErrors,
        } = await lift.inferMigrationSteps({
          migrationId: id,
          datamodel: datamodel,
          assumeToBeApplied: [],
          sourceConfig: datamodel,
        })

        if (stepErrors.length > 0) {
          throw stepErrors
        }

        const { errors } = await lift.applyMigration({
          force: true,
          migrationId: id,
          steps: datamodelSteps,
          sourceConfig: datamodel,
        })

        if (errors.length > 0) {
          throw errors
        }

        const progress = () =>
          lift.migrationProgess({
            migrationId: id,
            sourceConfig: datamodel,
          })

        while ((await progress()).status !== 'MigrationSuccess') {
          /* Just wait */
        }

        /* Release resource pool. */
        this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

        const instance: DBInstance = {
          url: dbFile,
          cwd: tmpDir,
          datamodel: datamodel,
        }

        /**
         * If there's a waiter in line allocate an instance, otherwise make it idle.
         */
        if (this.waiters.length > 0) {
          const [waiter, ...remainingWaiters] = this.waiters
          this.waiters = remainingWaiters
          this.dbs.busy = [...this.dbs.busy, instance]
          waiter.allocate(instance)
        } else {
          this.dbs.idle = [...this.dbs.idle, instance]
        }
      } catch (err) {
        throw err
      }
    }

    /**
     * Run the encapsulated funciton.
     *
     * @param fn
     */
    async run<T>(fn: (db: DBInstance) => Promise<T>) {
      const db = await this.getDBInstance()
      try {
        const result = await fn(db)
        await this.releaseDBInstance(db)
        return result
      } catch (e) {
        await this.releaseDBInstance(db)
        throw e
      }
    }

    /**
     * Returns a Photon instance from the pool once made available, and cleans
     * and populates the instance if necessary.
     *
     * The returned Photon instance includes the db identifier used for
     * the database releasing.
     *
     * @param opts
     */
    async getDBInstance(): Promise<DBInstance> {
      /**
       * Check whether any of the instances is available and creates a new one
       * if the pool limit is not yet reached, otherwise waits for the available
       * instance.
       */
      if (this.dbs.idle.length > 0) {
        /* If available take a resource and return it right away. */
        const [db, ...remaining] = this.dbs.idle
        this.dbs.idle = remaining
        this.dbs.busy = [...this.dbs.busy, db]

        return db
      } else {
        /* Add to the line. */
        const waiter = new Waiter()
        this.waiters = [...this.waiters, waiter]

        /**
         * If full capacity is not yet reached create new instance.
         * Otherwise, throw an error.
         */
        if (
          this.dbs.idle.length +
            this.dbs.busy.length +
            this.dbs.booting.length <
          this.capacity
        ) {
          this.createDBInstance()
        } else {
          throw new Error(`You've reached the upper limit of instances.`)
        }

        return waiter.wait()
      }
    }

    /**
     * Makes the instance available.
     */
    async releaseDBInstance(db: DBInstance): Promise<void> {
      /**
       * Find the busy instance, remove that instance and give
       * it to the next in line.
       */
      const instance = this.dbs.busy.find(bdb => bdb.cwd === db.cwd)!
      this.dbs.busy = this.dbs.busy.filter(bdb => bdb.cwd !== db.cwd)

      /* Allocate to the next waiter in line or make idle. */
      // if (this.waiters.length > 0) {
      //   const [waiter, ...remainingWaiters] = this.waiters
      //   this.waiters = remainingWaiters
      //   waiter.allocate(instance)
      // } else {
      //   this.dbs.idle = [...this.dbs.idle, instance]
      // }

      this.dbs.idle = [...this.dbs.idle, instance]
    }

    /**
     * Drains the pool by deleting all instances.
     */
    async drain(): Promise<void> {
      // TODO
    }
  }

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void = () => {}

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
