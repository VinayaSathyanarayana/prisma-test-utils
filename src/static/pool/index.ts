export { getPostgreSQLPool, PostgreSQLPoolOptions } from './dbs/postgres'
export { getMySQLPool, MySQLPoolOptions } from './dbs/mysql'
export { getSQLitePool, SQLitePoolOptions, getTmpSQLiteDB } from './dbs/sqlite'
