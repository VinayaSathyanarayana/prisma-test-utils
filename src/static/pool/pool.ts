import { Pool, DBInstance } from '../types'

export type PoolOptions = {
  max: number
}

export abstract class InternalPool implements Pool {
  protected dbs: {
    booting: string[]
    busy: DBInstance[]
  }
  protected waiters: Waiter[]
  protected capacity: number

  constructor(options: PoolOptions) {
    this.dbs = {
      booting: [],
      busy: [],
    }
    this.waiters = []
    this.capacity = options.max
  }

  /**
   * A required method used to create new databases in the pool.
   */
  protected abstract async createDBInstance(id?: string): Promise<DBInstance>

  /**
   * A required methdo used to delete databases in the pool.
   *
   * @param instance
   */
  protected abstract async deleteDBInstance(
    instance?: DBInstance,
  ): Promise<void>

  /**
   * Creates a new DBInstance and occupies the space in the pool.
   */
  public async getDBInstance(): Promise<DBInstance> {
    /**
     * If full capacity is not yet reached create new instance.
     * Otherwise, return a waiter which will wait until there's
     * an available space.
     */
    if (this.dbs.busy.length + this.dbs.booting.length < this.capacity) {
      /* Generates unique DB identifier. */
      const id = Math.random()
        .toString(36)
        .slice(2)

      /* Creates a new DBInstance. */
      this.dbs.booting = this.dbs.booting.concat(id)
      const dbInstance = await this.createDBInstance(id)
      this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

      this.dbs.busy = this.dbs.busy.concat(dbInstance)

      /**
       * If there's a waiter in the line it first gets the instnace.
       * A new waiter is created for the current request.
       *
       * If there are no waiters in the line, we return the DB instance.
       */
      const [waiter, ...remainingWaiters] = this.waiters
      if (waiter) {
        /**
         * Gives an instances to existing waiter and
         * returns a new waiter.
         */
        waiter.allocate(dbInstance)

        const newWaiter = new Waiter()
        this.waiters = [...remainingWaiters, newWaiter]

        return waiter.wait()
      } else {
        return dbInstance
      }
    } else {
      /* Add to the line. */
      const waiter = new Waiter()
      this.waiters = [...this.waiters, waiter]

      return waiter.wait()
    }
  }

  /**
   * Releases a db isntance in the pool and triggers the creation
   * of new instance if there are waiters.
   *
   * @param db
   */
  public async releaseDBInstance(db: DBInstance): Promise<void> {
    const instance = this.dbs.busy.find(bdb => bdb.url === db.url)!
    try {
      /* Finds the busy instance and releases it. */
      await this.deleteDBInstance(instance)
      this.dbs.busy = this.dbs.busy.filter(bdb => bdb.url !== db.url)

      /* Triggers the creation of new instance if there's a waiter for it. */
      if (this.waiters.length > 0) this.getDBInstance()
    } catch (err) {
      throw err
    }
  }
  /**
   * A function wrapper which makes sure that the db instance is
   * correctly allocated prior to execution of the function, and correctly
   * released after the execution has finished.
   *
   * @param fn
   */
  public async run<T>(fn: (db: DBInstance) => Promise<T>): Promise<T> {
    const db = await this.getDBInstance()
    try {
      const result = await fn(db)
      await this.releaseDBInstance(db)
      return result
    } catch (e) {
      await this.releaseDBInstance(db)
      throw e
    }
  }

  /**
   * Releases all remaining instances in the pool.
   */
  public async drain(): Promise<void> {
    const actions = this.dbs.busy.map(i => this.releaseDBInstance(i))
    try {
      await Promise.all(actions)
    } catch (err) {
      throw err
    }
  }
}

/* Pools specific to a DB type. */

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void = () => {}

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
