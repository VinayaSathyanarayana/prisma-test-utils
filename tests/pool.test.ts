import pg from 'pg'

import { InternalPool } from '../src/static/pool/pool'
import { DBInstance } from '../src/static/types'

describe('pool:', () => {
  test('manages dbs in the pool limit', async () => {
    class TestPool extends InternalPool {
      async createDBInstance(id: string) {
        return { url: '', cwd: '', datamodel: '' }
      }
      async deleteDBInstance(instance: DBInstance): Promise<void> {}
    }

    const pool = new TestPool({
      max: 1,
    })

    let released = false

    const instance = pool.getDBInstance().then(async instance => {
      expect(released).toBeFalsy()

      await pool.releaseDBInstance(instance)
      released = true
    })
    const waiter = pool.getDBInstance().then(instance => {
      expect(released).toBeTruthy()
    })

    await Promise.all([instance, waiter])
  })

  test('run allocates and releases instance', async () => {
    class TestPool extends InternalPool {
      async createDBInstance(id: string) {
        return { url: '', cwd: '', datamodel: '' }
      }
      async deleteDBInstance(instance: DBInstance): Promise<void> {}
    }

    const pool = new TestPool({
      max: 1,
    })

    await pool.run(async instance => {})

    /* Tests */
  })

  test('run allocates and releases instance on error', async () => {
    class TestPool extends InternalPool {
      async createDBInstance(id: string) {
        return { url: '', cwd: '', datamodel: '' }
      }
      async deleteDBInstance(instance: DBInstance): Promise<void> {}
    }

    const pool = new TestPool({
      max: 1,
    })

    await pool.run(async instance => {
      throw new Error('PASS')
    })

    /* Tests */
  })

  test('drains the dbs', async () => {
    /* Test pool instance. */
    class TestPool extends InternalPool {
      async createDBInstance(id: string) {
        return { url: '', cwd: '', datamodel: '' }
      }
      async deleteDBInstance(instance: DBInstance): Promise<void> {}
    }

    const pool = new TestPool({
      max: 1,
    })

    /* Create N instances. */
    const numberOfNewSchemas = Math.round(Math.random() * 30)
    await Promise.all(
      Array.from({ length: numberOfNewSchemas }, () => pool.getDBInstance()),
    )

    /* Drain pool. */
    await pool.drain()

    /* Tests. */
  })
})
