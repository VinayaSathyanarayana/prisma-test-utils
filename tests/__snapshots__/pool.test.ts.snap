// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pool: manages dbs in the pool limit 1`] = `
"datasource db {
  provider = \\"postgresql\\"
  url      = \\"postgres://prisma:prisma@127.0.0.1:5433/prisma?schema=prisma-test-utils-0quoet9gzk7d\\"
}

model User {
  id       String  @default(cuid()) @id
  name     String
  email    String
  isActive Boolean
}"
`;

exports[`pool: manages dbs in the pool limit 2`] = `
"datasource db {
  provider = \\"postgresql\\"
  url      = \\"postgres://prisma:prisma@127.0.0.1:5433/prisma?schema=prisma-test-utils-0quoet9gzk7d\\"
}

model User {
  id       String  @default(cuid()) @id
  name     String
  email    String
  isActive Boolean
}"
`;

exports[`pool: run allocates and releases instance 1`] = `
"datasource db {
  provider = \\"postgresql\\"
  url      = \\"postgres://prisma:prisma@127.0.0.1:5433/prisma?schema=prisma-test-utils-165fxnh84tyh\\"
}

model User {
  id       String  @default(cuid()) @id
  name     String
  email    String
  isActive Boolean
}"
`;

exports[`pool: run allocates and releases instance on error 1`] = `
"datasource db {
  provider = \\"postgresql\\"
  url      = \\"postgres://prisma:prisma@127.0.0.1:5433/prisma?schema=prisma-test-utils-klui39uk5n\\"
}

model User {
  id       String  @default(cuid()) @id
  name     String
  email    String
  isActive Boolean
}"
`;
